File: C# Fundamentals/C# OOP Basics/07. Encapsulation/Ex/P04_ShoppingSpree/Person.cs
using System;
using System.Collections.Generic;
using System.Text;

public class Person
{
    private string name;
    private decimal money;
    private List<Product> bag;

    public Person(string name, decimal money)
    {
        this.Name = name;
        this.Money = money;
        this.bag = new List<Product>();
    }

    public void AddProduct(Product product)
    {
        this.bag.Add(product);
    }

    public List<Product> Bag
    {
        get
        {
            return this.bag;
        }
    }
    public string Name
    {
        get
        {
            return this.name;
        }
        private set
        {
            if (string.IsNullOrWhiteSpace(value))
            {
                throw new ArgumentException("Name cannot be empty");
            }
            this.name = value;
        }
    }
    public decimal Money
    {
        get
        {
            return this.money;
        }
        set
        {
            if(value < 0)
            {
                throw new ArgumentException("Money cannot be negative");
            }
            this.money = value;
        }
    }
}

File: Java/01.Advanced/02.OOP/07.Encapsulation/Ex/src/t_03_shopping_spree/Person.java
package t_03_shopping_spree;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

public class Person {
    private String name;
    private double money;
    private final List<Product> products;

    public Person(String name, double money) {
        this.setName(name);
        this.setMoney(money);
        products = new ArrayList<>(5);
    }

    public void buyProduct(Product product) {
        if (product.getCost() > this.money) {
            throw new IllegalArgumentException(String.format("%s can't afford %s", this.name, product.getName()));
        } else {
            money -= product.getCost();
            products.add(product);
            System.out.printf("%s bought %s%n", name, product.getName());
        }
    }

    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder(26);

        if (products.isEmpty()) {
            sb.append(name);
            sb.append(" – Nothing bought");
        } else {
            sb.append(name);
            sb.append(" - ");
            sb.append(products.stream().map(Product::toString).collect(Collectors.joining(", ")));
        }
        return sb.toString();
    }

    public String getName() {
        return name;
    }

    private void setName(String name) {
        if (name == null || name.trim().isEmpty()) {
            throw new IllegalArgumentException("Name cannot be empty");
        }
        this.name = name;
    }

    private void setMoney(double money) {
        if (money < 0) {
            throw new IllegalArgumentException("Money cannot be negative");
        }
        this.money = money;
    }
}
